### PR TITLE
[view-transitions] Support view transitions on top layer elements

### PR DESCRIPTION
**Pull request by mattwoodrow**
#### 2e0c7a1325e77dfe9a67c60778af5ce27efa4c78
<pre>
[view-transitions] Support view transitions on top layer elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=265169">https://bugs.webkit.org/show_bug.cgi?id=265169</a>
&lt;<a href="https://rdar.apple.com/118667007">rdar://118667007</a>&gt;

Reviewed by Tim Nguyen.

If the captured element is the document element, then the captured images should be
of the RenderView instead.

We already adjust the sizing to that of the view, this changes the actual snapshots to
happen from the view so that the top layer is included.

Changes the layer parenting for the new snapshot so that the RenderView&apos;s layer gets added
to the ::view-transition-new(pseudo), and the ::view-transition root pseudo get directly
parented in place of the RenderView.

The render tree for this situation is this:

Content-root
     |
 RenderView
   /     \
&lt;html&gt;   ::view-transition
            \
           ::view-transition-group(root)
             |
           ::view-transition-image-pair(root)
             |
           ::view-transition-new(root)

We want to build a GraphicsLayer tree where the RenderView and all its contents (except the view-transition
tree) are attached to the ::view-transition-new.

Content-root
     |
::view-transition
     |
::view-transition-group(root)
     |
::view-transition-image-pair(root)
     |
::view-transition-new(root)
     |
 RenderView
     |
   &lt;html&gt;

This happens by:

RenderLayerCompositor::collectViewTransitionNewContentLayers detects when the captured element is the &lt;html&gt;,
and adjusts to reparent the RenderView instead. This ensures the &apos;top layer&apos; is included, as required.

RenderLayerCompositor::updateBackingAndHierarchy skips attaching the ::view-transition&apos;s GraphicsLayer
to the parent obtained via the RenderLayer hierarchy (the RenderView&apos;s layer).

RenderLayerBacking::childForSuperlayers detects when it&apos;s a RenderView and will already be attached to
the ::view-transition-new, and forwards the request through to the ::view-transition&apos;s RenderLayer,
so that ::view-transition gets attached to the parent of the RenderView instead.

Adds a new test for the ::view-transition-old(root) case.

* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/dialog-in-rtl-iframe-old-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/dialog-in-rtl-iframe-old-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/dialog-in-rtl-iframe-old.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/support/dialog-in-rtl-iframe-child-old.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::activeViewTransitionCapturedDocumentElement const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::snapshotElementVisualOverflowClippedToViewport):
(WebCore::ViewTransition::documentElementIsCaptured const):
* Source/WebCore/dom/ViewTransition.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::isViewTransitionRoot const):
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintList):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::childForSuperlayers const):
(WebCore::RenderLayerBacking::childForSuperlayersExcludingViewTransitions const):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::collectViewTransitionNewContentLayers):
(WebCore::RenderLayerCompositor::updateBackingAndHierarchy):

Canonical link: <a href="https://commits.webkit.org/278256@main">https://commits.webkit.org/278256@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/786d2f6d7a691a135bd606dec072f26a3510fd8d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50023 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29312 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2306 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53266 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/700 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52324 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35447 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/269 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/40796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52121 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/26920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43023 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/21928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/24352 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/263 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8393 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/46396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/310 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54848 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25117 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/48190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/26375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43223 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10966 "Built successfully and passed tests") | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-10-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26104 "Built successfully") | | |

### CODE REVIEW

File: LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/support/dialog-in-rtl-iframe-child-old.html [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      }
+
+    </style>
+    <script>
+      function show() {
+        document.getElementById('dialog').showModal();
+      }
+      function startTransition() {
+        show();
+        window.transition = document.startViewTransition(() => {
```
Comment (nt1m) [inline LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/support/dialog-in-rtl-iframe-child-old.html:35]:
```suggestion
        document.startViewTransition(() => {
```

Can probably be omitted

File: Source/WebCore/rendering/RenderLayerBacking.cpp [hunk 2]
```diff
@@ -3319,6 +3319,21 @@ GraphicsLayer* RenderLayerBacking::parentForSublayers() const
 }
 
 GraphicsLayer* RenderLayerBacking::childForSuperlayers() const
+{
+    if (m_owningLayer.isRenderViewLayer()) {
+        // If the document element is captured, then the RenderView's layer will get attached
+        // into the view-transition tree, and we instead want to attach the root of the VT tree to our ancestor.
+        Ref document = m_owningLayer.renderer().document();
+        if (RefPtr activeViewTransition = document->activeViewTransition(); activeViewTransition && activeViewTransition->documentElementIsCaptured()) {
+            RenderLayer* viewTransitionRoot = m_owningLayer.lastChild();
```
Comment (nt1m) [inline Source/WebCore/rendering/RenderLayerBacking.cpp]: auto* or smart pointer

**Review by nt1m — COMMENTED** [review]
I'll take a deeper look later.

File: LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/support/dialog-in-rtl-iframe-child-old.html (hunk 1 above)
Comment (mattwoodrow) [reply LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/support/dialog-in-rtl-iframe-child-old.html:35]: The outer document uses this to decide when to take the snapshot.

**Review by nt1m — APPROVED** [review]
LGTM, Can you add clarification in the commit message about what we discussed?

**Pull request merged**